[feat] : 공고 현황 페이지 공고글 필터링 기능 & 무한스크롤 기능 구현

rktdnjs opened the pull request:
## PR 타입(하나 이상의 PR 타입을 선택해주세요)

- [x] 기능 추가
- [ ] 기능 삭제
- [ ] 버그 수정
- [ ] CSS등 UI 수정
- [x] 의존성, 환경 변수, 빌드 관련 코드 업데이트

## 작업 사항

- `FilterBtn` 구현(기존에 쓰던 Button에 영향을 주지 않기 위해 별도 구현)
- `PostListPage - FilterForm - FilterBtn` 사이의 데이터 흐름
- 최상단 `PostListPage`에서 데이터 페칭
- `FilterForm`에서 필터 버튼 클릭 - 필터값 `PostListPage`로 전달
- `PostListPage`에서는 필터값을 바탕으로 페칭한 데이터 필터링 하여 렌더링 처리
- 데이터페칭 코드 추가후에 컴포넌트간 데이터 흐름 디버깅 필요할 것으로 예상

![chrome-capture-2023-9-13](https://github.com/Step3-kakao-tech-campus/Team12_FE/assets/67001905/aa189bb6-a966-49c3-b475-86916088665d)


## 관련 이슈

#34

rktdnjs committed:
[feat] : 공고 현황 페이지 필터처리용 버튼 구현

rktdnjs committed:
[feat] : 공고 현황 페이지 필터링 기능 구현

- PostListPage - FilterForm - FilterBtn 사이의 데이터 흐름
- 최상단 PostListPage에서 데이터 페칭
- FilterForm에서 필터 버튼 클릭 - 필터값 PostListPage로 전달
- PostListPage에서는 필터값을 바탕으로 페칭한 데이터 필터링 하여 렌더링 처리
- 데이터페칭 코드 추가후에 컴포넌트간 데이터 흐름 디버깅 필요할 것으로 예상

rktdnjs committed:
[refactor] : Cards 컴포넌트  key attribute 추가

rktdnjs commented:
## 작업 사항

- 무한스크롤 감지를 위한 `react-intersection-observer` 설치 및 추가
- 공고 현황 페이지에서 맨 하단에 다다랐을 때 `useInView`에 의해 요소 감지, 해당 로직을 통해 추가 페이지 렌더링
- 공고글을 가져오기 위한 `post.js` API 처리 로직 추가(정상 동작하는지는 API 배포 후 확인 or 모킹을 통해 확인 필요)
- `react-intersection-observer useInView` 훅 & `react-query useInfiniteQuery` 훅 활용
- 스크롤 감지 = `useInView` / 데이터 페칭 = `useInfiniteQuery`
- `useInView`에서 제공하는 `ref`를 페이지 요소로 걸어두고 `inView`값이 노출되면 추가 데이터를 불러오는 구조
- 초회에 `offset`을 아무것도 주지 않고 요청하여 받아오는 데이터에서 마지막 공고를 id값을 이용하여 pageOffset으로 이용
- 만약 마지막으로 받아왔던 데이터에 `last` 값이 `true`일 경우, 더이상 데이터를 불러오지 않도록 params로 `null`을 리턴
- eslint로 인해 오류로 잡는 부분이 많아 현재는 페이지가 돌아가도록 잡다한 코드가 많은 상태
- 백엔드 API연결 이후에 전체적으로 다시 리팩토링 진행 예정

## 관련 이슈

#34

rktdnjs committed:
[conf] : 무한스크롤 감지를 위한 react-intersection-observer 라이브러리 추가

rktdnjs committed:
[feat] : 스크롤 하단부 감지 추가

rktdnjs committed:
[feat] : 공고글 현황 API 로직 추가

baegyeong commented on src/pages/PostListPage.jsx:
```diff
@@ -62,31 +71,56 @@ const PostListPage = () => {
+    if (inView) {
+      console.log('다음 페이지 요청 & inView : ', inView);
+    }
+  }, [inView]);
```
`useInView` 잘 적용해주셨네요 !!! 👍🏻

rktdnjs committed:
[feat] : 공고 현황 페이지 무한스크롤 구현(API 연결 확인 x)

- react-intersection-observer useInView 훅 & react-query useInfiniteQuery 훅 활용
- 스크롤 감지 = useInView / 데이터 페칭 = useInfiniteQuery
- useInView에서 제공하는 ref를 페이지 요소로 걸어두고 inView값이 노출되면 추가 데이터를 불러오는 구조
- 초회에 offset을 아무것도 주지 않고 요청하여 받아오는 데이터에서 마지막 공고를 id값을 이용하여 pageOffset으로 이용
- 만약 마지막으로 받아왔던 데이터에 last 값이 true일 경우, 더이상 데이터를 불러오지 않도록 params로 null을 리턴
- eslint로 인해 오류로 잡는 부분이 많아 현재는 페이지가 돌아가도록 코드가 많은 상태
- 백엔드 API연결 이후에 전체적으로 다시 리팩토링 진행 예정

rktdnjs commented:
현재 `index.js` 내부에 구현된 `axios instance`를 사용하는 로직을 가져다가 사용할 경우에 에러가 발생하는 것 같습니다.
![image](https://github.com/Step3-kakao-tech-campus/Team12_FE/assets/67001905/988ef612-5b7f-499a-987c-94233f810f8b)
제 생각에는, 현재 `index.js`에서 `instance`를 만드는 과정에서 정의된 `REACT_APP_API_URL` 가 따로 정의되어 있지 않아 해당 데이터를 읽어와서 인스턴스를 생성하는 과정에서 에러를 내뱉는 것 같다고 판단됩니다!
따라서 이 부분은 차후 백엔드 API base URL을 `.env` 파일에 추가해주면 해결되지 않을까, 생각해봅니다..!

```
(수정) REACT_APP_API_URL가 문제가 아니라, 제가 최근에 수정했던 apis/index.js에서 사용한 useNavigate()가 문제였습니다...
사실 예전에도 비슷한 문제로 골머리 앓았었는데, 다시 깨닫게 되네요. 
리액트 hook들은 함수형 컴포넌트 내부에서 사용되어야 하는데, 그러지 않아 발생하는 문제였습니다! 
따라서 문제가 되는 부분을 수정하였습니다.
```

rktdnjs committed:
[feat] : react-query 사용을 위한 QueryClientProvider 추가 + ReactStrictMode 제거

rktdnjs committed:
[fix] : invalid API call - api/index.js useNavigate 제거
